### DISCUSSION TITLE
ConnectionStatusChanged is not fired after server restart

### DISCUSSION THREAD

**Discussion by wallaceturner** [ravendb/ravendb#17474] [General]
Hi Support we are observing some intermittent behavior in the changes API using 5.4.105 (client & server)

We have a windows process that has wired up `ConnectionStatusChanged`

```
store.Changes().ConnectionStatusChanged += (object sender, EventArgs e) => Log.Info($"ConnectionStatusChanged Connected:{((IDatabaseChanges)sender).Connected}");
```

If we start out process then restart ravendb we consistently see the expected 2 logs statements
```
ConnectionStatusChanged Connected:False
...
ConnectionStatusChanged Connected:True
```
All good so far however we have run into an intermittent issue where the reconnect event is seemingly not fired if ravendb has been running for some time and we restart it - the `Connected:False `is logged but not the subsequent `Connected:True`

Has their been any changes/fixes in this component since 105? We reviewed [your change log](https://ravendb.net/docs/article-page/5.4/Csharp/start/whats-new/1) it doesnt appear so

We don't have a stable recreate as its intermittent but can also report that multiple subsequent restarts of ravendb do not then trigger *any logging* - its as if the event was never wired

**Comment by wallaceturner** [comment]
I have a fairly stable recreate now - i have a simple console app connecting to a ravendb instance

```
var store = new DocumentStore { Database = "...", Urls = new[] { "https://....fglau.com:8080" } };
store.Certificate = RavenDbExtensions.GetCert();
store.Initialize();
store.Changes().ConnectionStatusChanged += (sender, args) => Console.WriteLine($"Connected:{((IDatabaseChanges)sender).Connected}");
```

I have this Console app running on 2 servers: 1 on the same network as the ravendb instance and the other on the ravendb instance itself.

After restarting ravendb a few times (wait a few mins between restarts) eventually the Console app running over the network fails:  (i.e. the console app **not** running on same server as ravendb) 



![image](https://github.com/ravendb/ravendb/assets/2849980/1d5b9c46-5e2e-4d38-b29b-2a38532198ba)

Are you able to provide any more info as to why this might be happening? Once its in this state it does not reconnect (and changes API no longer works; subscribers do not get changes)

antivirus (symantec) was disabled for the duration of the test.

**Comment by wallaceturner** [comment]
this is the state of `store.Changes()` when it is hung

![image](https://github.com/ravendb/ravendb/assets/2849980/5181b4c1-e97f-4a9a-b1d6-665f5d5345cd)

**Comment by wallaceturner** [comment]
I have investigated this today and the culprit is the call to `ConnectAsync`
https://github.com/ravendb/ravendb/blob/3e05b49913a0bbff9fc14019dccfdb1409a5e4d7/src/Raven.Client/Documents/Changes/DatabaseChanges.cs#L549

This call hangs 'sometimes' - the recreate is to connect to a server from win10 or windows 2012 (win 11 seems ok) and STOP the ravendb service. Adding console statements to your code i verified the reconnect attempts - interestingly the hang does not occur until ravendb is started. If you attempt the recreate do not simply restart ravendb - stop, wait 1min then start

Proposed fix - add a timeout

https://github.com/wallaceturner/ravendb/blob/7146a594dcf04360e96cc7707b95a36be8ad1805/src/Raven.Client/Documents/Changes/DatabaseChanges.cs#L548-L549

**Comment by garayx** [comment]
Hi

Can you upgrade both client & server to the latest RavenDB 5.4.111 and see if this fixes the issue ?

**Reply by wallaceturner** [reply]
Hi Egor thanks for the reply - we can bump the version in the fullness of time but its on a quasi-prod server so it wont be quick
Is there reason to believe this may be fixed by 5.4.111? the code looks the same (as does v6) - i.e. there is still a call to `ConnectAsync` that does not have a timeout

**Reply by wallaceturner** [reply]
hi Egor i'm able to recreate this in `5.4.111` too

**Comment by wallaceturner** [comment]
Hi Egor is it possible for you guys to review the fix I submitted for this issue and consider adding it in the next 5.4.x release?

**Comment by garayx** [comment]
hi,

I tried to repro the issue using the code you sent, but it works. 

I used RavenDB Instance on cloud.ravendb.net and 2 Console app running on my local machine with this code:

```
     using var store = new DocumentStore()
            {
                // Define the cluster node URLs (required)
                Urls = new[] { "http://52.145.11.1:8080/", 
                    /*some additional nodes of this cluster*/ },
                // Define a default database (optional)
                Database = "Data",


                // Initialize the Document Store
            }.Initialize();



            store.Changes().ConnectionStatusChanged += (sender, args) =>
            {
                Console.WriteLine($"Connected:{((IDatabaseChanges)sender).Connected}");


                Console.WriteLine();
            };






            Thread.Sleep(
                int.MaxValue);
```

I tried to wait 1 minute and, immediately restart both apps were reconnecting after start of the server.

![image](https://github.com/ravendb/ravendb/assets/16625387/ac857f8c-0f2d-47de-822e-9679c3283d27)


Can you send complete sample app that reproduce the issue ? or preferable a failing unit test: https://ravendb.net/docs/article-page/6.0/csharp/start/test-driver#complete-example

**Comment by wallaceturner** [comment]
host/client: windows 10
server: windows server 2012

here is a visual recreate:

https://www.youtube.com/watch?v=DcDYG-PjKIk

here is a video showing the hung ravendb task on the client

https://www.youtube.com/watch?v=GopDU6a93co


My summary is that you need to add a timeout to the `ConnectAsyncCore` - 

![image](https://github.com/ravendb/ravendb/assets/2849980/3ac53c42-cec4-4634-8e50-1d8b4cd22830)

**Reply by garayx** [reply]
Hi

I will try to check on Windows Server 2012, 

meanwhile, is there a chance you can [submit a Pull Request](https://github.com/ravendb/ravendb/blob/v5.4/CONTRIBUTING.md#submitting-a-pull-request) from you branch (https://github.com/wallaceturner/ravendb/tree/discussions-17474) ?

**Reply by wallaceturner** [reply]
hi egor can you please create a raven issue i can link to in the PR (as per your PR guidelines)

**Reply by garayx** [reply]
https://issues.hibernatingrhinos.com/issue/RavenDB-21692

**Reply by wallaceturner** [reply]
https://github.com/ravendb/ravendb/pull/17695
